HpToolsLauncher/Runners/FileSystemTestsRunner.cs

```diff
@@ -1,549 +1,578 @@
+                        };
+                    }
+
+                    //get the original source for this test, for grouping tests under test classes
+                    runResult.TestGroup = test.TestGroup;
+
+                    activeRunDesc.TestRuns.Add(runResult);
+
+                    //if fail was terminated before this step, continue
+                    if (runResult.TestState != TestState.Failed)
+                    {
+                        if (runResult.TestState != TestState.Error)
+                        {
+                            Helper.GetTestStateFromReport(runResult);
+                        }
+                        else
+                        {
+                            if (string.IsNullOrEmpty(runResult.ErrorDesc))
+                            {
+                                runResult.ErrorDesc = RunCancelled() ? Resources.ExceptionUserCancelled : Resources.ExceptionExternalProcess;
+                            }
+                            runResult.ReportLocation = null;
+                            runResult.TestState = TestState.Error;
+                        }
+                    }
+
+                    if (runResult.TestState == TestState.Passed && runResult.HasWarnings)
+                    {
+                        runResult.TestState = TestState.Warning;
+                        ConsoleWriter.WriteLine(Resources.FsRunnerTestDoneWarnings);
+                    }
+                    else
+                    {
+                        ConsoleWriter.WriteLine(string.Format(Resources.FsRunnerTestDone, runResult.TestState));
+                        if (runResult.TestState == TestState.Error)
+                        {
+                            ConsoleWriter.WriteErrLine(runResult.ErrorDesc);
+                        }
+                    }
+
+                    UpdateCounters(runResult.TestState);
+                    var testTotalTime = (DateTime.Now - testStart).TotalSeconds;
+                    ConsoleWriter.WriteLine(DateTime.Now.ToString(Launcher.DateFormat) + " Test completed in " + ((int)Math.Ceiling(testTotalTime)).ToString() + " seconds: " + runResult.TestPath);
+                    if (!string.IsNullOrWhiteSpace(runResult.ReportLocation))
+                    {
+                        ConsoleWriter.WriteLine(DateTime.Now.ToString(Launcher.DateFormat) + " Test report is generated at: " + runResult.ReportLocation + NEW_LINE_AND_DASH_SEPARATOR);
+                    }
+
+                    if (_cancelRunOnFailure && (idx < _tests.Count - 1) && (runResult.TestState == TestState.Failed || runResult.TestState == TestState.Error))
+                    {
+                        Console.WriteLine(Resources.FileSystemTestsRunner_Run_Auto_Cancelled + NEW_LINE_AND_DASH_SEPARATOR);
+                        break;
+                    }
+                }
+
+                totalTime = (DateTime.Now - start).TotalSeconds;
+
+                for (++idx; idx < _tests.Count; idx++)

```

Comment:
Suggest to add code comments here like:
```cs
// skip the rest test
for (++idx; ...
```

Reply:
Ok, it sounds better to use one foreach loop and outer skipRemainingTests.